File: app/src/main/java/com/lifetrace/execute/domain/project/ExecutionProject.kt
package com.lifetrace.execute.domain.project

enum class ExecutionProjectStatus(val wireValue: String) {
    ACTIVE("active"),
    PAUSED("paused"),
    COMPLETED("completed"),
    ARCHIVED("archived");

    companion object {
        fun fromWire(value: String): ExecutionProjectStatus =
            entries.firstOrNull { it.wireValue == value } ?: ACTIVE
    }
}

data class ExecutionProject(
    val id: String,
    val userId: String,
    val name: String,
    val description: String? = null,
    val status: ExecutionProjectStatus = ExecutionProjectStatus.ACTIVE,
    val dueAt: String? = null,
    val createdAt: String,
    val updatedAt: String,
    val localVersion: Long,
    val serverVersion: String? = null,
    val modifiedByDevice: String? = null,
)

data class ProjectProgress(
    val project: ExecutionProject,
    val taskCount: Int,
    val completedTaskCount: Int,
) {
    val completionRatio: Float
        get() = if (taskCount == 0) 0f else completedTaskCount.toFloat() / taskCount.toFloat()
}

File: flutter_app/lib/domain/project/execution_project.dart
enum ExecutionProjectStatus {
  active('active'),
  paused('paused'),
  completed('completed'),
  archived('archived');

  const ExecutionProjectStatus(this.wireValue);
  final String wireValue;

  static ExecutionProjectStatus fromWire(String value) =>
      ExecutionProjectStatus.values.firstWhere(
        (status) => status.wireValue == value,
        orElse: () => ExecutionProjectStatus.active,
      );
}

class ExecutionProject {
  const ExecutionProject({
    required this.id,
    required this.userId,
    required this.title,
    required this.createdAt,
    required this.updatedAt,
    required this.localVersion,
    this.description,
    this.status = ExecutionProjectStatus.active,
    this.startAt,
    this.dueAt,
    this.serverVersion,
    this.modifiedByDevice,
  });

  final String id;
  final String userId;
  final String title;
  final String? description;
  final ExecutionProjectStatus status;
  final String? startAt;
  final String? dueAt;
  final String createdAt;
  final String updatedAt;
  final int localVersion;
  final String? serverVersion;
  final String? modifiedByDevice;

  bool get isArchived => status == ExecutionProjectStatus.archived;
  bool get isCompleted => status == ExecutionProjectStatus.completed;

  ExecutionProject copyWith({
    String? title,
    String? description,
    ExecutionProjectStatus? status,
    String? startAt,
    String? dueAt,
    String? updatedAt,
    int? localVersion,
    String? serverVersion,
    String? modifiedByDevice,
    bool clearDescription = false,
    bool clearStartAt = false,
    bool clearDueAt = false,
  }) {
    return ExecutionProject(
      id: id,
      userId: userId,
      title: title ?? this.title,
      description: clearDescription ? null : description ?? this.description,
      status: status ?? this.status,
      startAt: clearStartAt ? null : startAt ?? this.startAt,
      dueAt: clearDueAt ? null : dueAt ?? this.dueAt,
      createdAt: createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      localVersion: localVersion ?? this.localVersion,
      serverVersion: serverVersion ?? this.serverVersion,
      modifiedByDevice: modifiedByDevice ?? this.modifiedByDevice,
    );
  }
}
